File: backend/app/risk.py
"""Risk engine ported from frontend/services/risk.js. Pure functions, no I/O."""

from __future__ import annotations

SIGNAL_KEYS = [
    "downward_trend",
    "missed_engagement",
    "low_projected_grade",
    "heavy_commitments",
    "missing_data",
]

SIGNAL_LABELS = {
    "downward_trend": "Downward trend",
    "missed_engagement": "Missed or low engagement",
    "low_projected_grade": "Low projected grade",
    "heavy_commitments": "Heavy outside commitments",
    "missing_data": "Missing data",
}

DEFAULT_THRESHOLDS = {
    "declineWeeks": 3,
    "missedAssignments": 2,
    "attendancePct": 70,
    "participationPct": 50,
    "projectedGrade": 60,
    "commitmentHours": 20,
    "missingWeeks": 2,
}

DEFAULT_ACTIVE = {k: True for k in SIGNAL_KEYS}
LEVELS = ["Not flagged", "Watch", "At risk", "High risk"]


def level_for(count: int) -> str:
    if count >= 3:
        return "High risk"
    return LEVELS[count] if 0 <= count < len(LEVELS) else "Not flagged"


def _round1(n):
    return None if n is None else round(n * 10) / 10


def _pct(earned, possible):
    return (100 * earned) / possible if possible and possible > 0 else None


def _get(obj, key, default=None):
    if obj is None:
        return default
    if isinstance(obj, dict):
        return obj.get(key, default)
    return getattr(obj, key, default)


def _decline_run(weekly) -> int:
    scored = [w for w in weekly if _get(w, "raw") is not None]
    run = 0
    for i in range(len(scored) - 1, 0, -1):
        curr = _get(scored[i], "raw")
        prev = _get(scored[i - 1], "raw")
        if curr is None or prev is None:
            break
        if curr < prev - 1e-9:
            run += 1
        else:
            break
    return run


def _missing_run(weekly) -> int:
    best = 0
    run = 0
    for w in weekly:
        if _get(w, "hasEntries"):
            run = 0
        else:
            run += 1
            best = max(best, run)
    return best


def _missed_assignments(weekly) -> int:
    count = 0
    for w in weekly:
        by_key = _get(w, "byKey", {}) or {}
        slot = by_key.get("homework")
        if slot and (slot.get("possible", 0) or 0) > 0 and (slot.get("earned", 0) or 0) == 0:
            count += 1
    return count


def _weeks_set(weekly) -> int:
    count = 0
    for w in weekly:
        by_key = _get(w, "byKey", {}) or {}
        slot = by_key.get("homework")
        if slot and (slot.get("possible", 0) or 0) > 0:
            count += 1
    return count


def _totals_for(weekly, key: str) -> dict:
    earned = 0
    possible = 0
    for w in weekly:
        by_key = _get(w, "byKey", {}) or {}
        slot = by_key.get(key)
        if not slot:
            continue
        earned += slot.get("earned", 0) or 0
        possible += slot.get("possible", 0) or 0
    return {"earned": earned, "possible": possible}


def evaluate_signals(*, weekly, hours: float = 0, thresholds: dict | None = None, active: dict | None = None) -> dict:
    t = {**DEFAULT_THRESHOLDS, **(thresholds or {})}
    weekly = list(weekly or [])
    scored = [w for w in weekly if _get(w, "raw") is not None]

    run = _decline_run(weekly)
    missed = _missed_assignments(weekly)
    set_count = _weeks_set(weekly)
    attendance = _totals_for(weekly, "attendance")
    attendance_pct = _pct(attendance["earned"], attendance["possible"])
    participation = _totals_for(weekly, "participation")
    participation_pct = _pct(participation["earned"], participation["possible"])
    projected = None
    if scored:
        vals = [_get(w, "raw") for w in scored]
        nums = [v for v in vals if v is not None]
        projected = (sum(nums) / len(nums)) if nums else None
    gap = _missing_run(weekly)

    maybe_reasons = [
        f"attendance {_round1(attendance_pct)}% is under {t['attendancePct']}%" if attendance_pct is not None and attendance_pct < t["attendancePct"] else None,
        f"{missed} assignments not handed in" if missed >= t["missedAssignments"] else None,
        f"participation {_round1(participation_pct)}% is under {t['participationPct']}%" if participation_pct is not None and participation_pct < t["participationPct"] else None,
    ]
    reasons: list[str] = [r for r in maybe_reasons if r]

    plural = lambda n: "" if n == 1 else "s"  # noqa: E731
    defs = {
        "downward_trend": {
            "on": run >= t["declineWeeks"],
            "summary": f"The weighted score has fallen {run} weeks running."
            if run >= t["declineWeeks"]
            else f"The score has fallen {run} week{plural(run)} running; the threshold is {t['declineWeeks']}.",
            "evidence": [
                {"label": f"Week {_get(w, 'week_number')}", "value": str(_round1(_get(w, 'raw')))}
                for w in scored[-6:]
            ],
        },
        "missed_engagement": {
            "on": len(reasons) > 0,
            "summary": f"Triggered by {'; '.join(reasons)}."
            if reasons
            else "Attendance, homework and participation are all above their thresholds.",
            "evidence": [
                {"label": "Assignments not handed in", "value": f"{missed} of {set_count}"},
                {"label": "Attendance to date", "value": "—" if attendance_pct is None else f"{_round1(attendance_pct)}%"},
                {"label": "Participation to date", "value": "—" if participation_pct is None else f"{_round1(participation_pct)}%"},
            ],
        },
        "low_projected_grade": {
            "on": projected is not None and projected < t["projectedGrade"],
            "summary": "No scores yet, so no projection."
            if projected is None
            else f"The score to date averages {_round1(projected)}, against a pass threshold of {t['projectedGrade']}.",
            "evidence": [
                {"label": "Weeks counted", "value": str(len(scored))},
                {"label": "Projected grade", "value": "—" if projected is None else str(_round1(projected))},
                {"label": "Pass threshold", "value": str(t["projectedGrade"])},
            ],
        },
        "heavy_commitments": {
            "on": (hours or 0) >= t["commitmentHours"],
            "summary": f"{_round1(hours)} weighted hours of work and care a week, at or above the {t['commitmentHours']}-hour threshold."
            if (hours or 0) >= t["commitmentHours"]
            else f"{_round1(hours)} weighted hours a week, under the {t['commitmentHours']}-hour threshold.",
            "evidence": [
                {"label": "Weighted hours this week", "value": str(_round1(hours))},
                {"label": "Threshold", "value": f"{t['commitmentHours']} hours"},
            ],
        },
        "missing_data": {
            "on": gap >= t["missingWeeks"],
            "summary": f"{gap} consecutive weeks with no entries at all."
            if gap >= t["missingWeeks"]
            else f"The longest gap is {gap} week{plural(gap)}; the threshold is {t['missingWeeks']}.",
            "evidence": [
                {"label": f"Week {_get(w, 'week_number')}", "value": "recorded" if _get(w, "hasEntries") else "no entries"}
                for w in weekly[-6:]
            ],
        },
    }

    active = active or {}
    signals = [
        {
            "key": key,
            "label": SIGNAL_LABELS[key],
            "active": active.get(key, True) is not False,
            "on": (active.get(key, True) is not False) and bool(defs[key]["on"]),
            "summary": defs[key]["summary"],
            "evidence": defs[key]["evidence"],
        }
        for key in SIGNAL_KEYS
    ]
    on = [s for s in signals if s["on"]]
    return {
        "signals": signals,
        "signals_on": [s["key"] for s in on],
        "level": level_for(len(on)),
        "oneAway": None if len(on) >= 3 else level_for(len(on) + 1),
        "metrics": {
            "declineRun": run,
            "missedAssignments": missed,
            "attendancePct": _round1(attendance_pct),
            "participationPct": _round1(participation_pct),
            "projected": _round1(projected),
            "hours": _round1(hours),
            "missingRun": gap,
        },
    }


def diff_state(prior: dict | None, current: dict) -> str:
    was = bool(prior and len(prior.get("signals_on", [])) > 0)
    is_now = len(current.get("signals_on", [])) > 0
    if is_now and not was:
        return "new"
    if is_now and was:
        return "still"
    if not is_now and was:
        return "cleared"
    return "quiet"

File: frontend/services/risk.js
// The risk engine. Pure functions over normalised weekly data — no I/O, no
// adapter detail. Each signal is on or off against an instructor-set threshold,
// and the level is the count of signals on. Nothing here is a blended number
// anyone has to interpret, and every signal carries the data that triggered it.

export const SIGNAL_KEYS = [
  'downward_trend',
  'missed_engagement',
  'low_projected_grade',
  'heavy_commitments',
  'missing_data',
];

export const SIGNAL_LABELS = {
  downward_trend: 'Downward trend',
  missed_engagement: 'Missed or low engagement',
  low_projected_grade: 'Low projected grade',
  heavy_commitments: 'Heavy outside commitments',
  missing_data: 'Missing data',
};

export const DEFAULT_THRESHOLDS = {
  declineWeeks: 3,
  missedAssignments: 2,
  attendancePct: 70,
  participationPct: 50,
  projectedGrade: 60,
  commitmentHours: 20,
  missingWeeks: 2,
};

export const DEFAULT_ACTIVE = Object.fromEntries(SIGNAL_KEYS.map((k) => [k, true]));

export const LEVELS = ['Not flagged', 'Watch', 'At risk', 'High risk'];

/** A student's level is the count of distinct signals currently on. */
export function levelFor(count) {
  return count >= 3 ? 'High risk' : LEVELS[count] || 'Not flagged';
}

const pct = (earned, possible) => (possible > 0 ? (100 * earned) / possible : null);
const round1 = (n) => (n == null ? null : Math.round(n * 10) / 10);

/** How many weeks in a row the raw score fell, counting back from the latest. */
function declineRun(weekly) {
  const scored = weekly.filter((w) => w.raw != null);
  let run = 0;
  for (let i = scored.length - 1; i > 0; i--) {
    if (scored[i].raw < scored[i - 1].raw - 1e-9) run += 1;
    else break;
  }
  return run;
}

/** The longest run of consecutive weeks with no entries at all. */
function missingRun(weekly) {
  let best = 0;
  let run = 0;
  for (const w of weekly) {
    if (w.hasEntries) run = 0;
    else {
      run += 1;
      if (run > best) best = run;
    }
  }
  return best;
}

/** Weeks where an assignment was set but nothing was handed in. */
function missedAssignments(weekly) {
  return weekly.filter((w) => {
    const slot = w.byKey && w.byKey.homework;
    return slot && slot.possible > 0 && slot.earned === 0;
  }).length;
}

function weeksSet(weekly) {
  return weekly.filter((w) => w.byKey && w.byKey.homework && w.byKey.homework.possible > 0).length;
}

function totalsFor(weekly, key) {
  let earned = 0;
  let possible = 0;
  for (const w of weekly) {
    const slot = w.byKey && w.byKey[key];
    if (!slot) continue;
    earned += slot.earned;
    possible += slot.possible;
  }
  return { earned, possible };
}

/**
 * Evaluate the five signals for one student up to and including a week.
 *
 * weekly: [{ week_number, raw, hasEntries, byKey: { [criterionKey]: {earned, possible} } }]
 *         — already limited to weeks the student was enrolled for.
 * hours:  weighted commitment hours in the week being evaluated.
 */
export function evaluateSignals({ weekly, hours = 0, thresholds = DEFAULT_THRESHOLDS, active = DEFAULT_ACTIVE }) {
  const t = { ...DEFAULT_THRESHOLDS, ...thresholds };
  const scored = weekly.filter((w) => w.raw != null);

  const run = declineRun(weekly);
  const missed = missedAssignments(weekly);
  const setCount = weeksSet(weekly);
  const attendance = totalsFor(weekly, 'attendance');
  const attendancePct = pct(attendance.earned, attendance.possible);
  const participation = totalsFor(weekly, 'participation');
  const participationPct = pct(participation.earned, participation.possible);
  const projected = scored.length ? scored.reduce((a, w) => a + w.raw, 0) / scored.length : null;
  const gap = missingRun(weekly);

  const reasons = [
    attendancePct != null && attendancePct < t.attendancePct
      ? `attendance ${round1(attendancePct)}% is under ${t.attendancePct}%` : null,
    missed >= t.missedAssignments ? `${missed} assignments not handed in` : null,
    participationPct != null && participationPct < t.participationPct
      ? `participation ${round1(participationPct)}% is under ${t.participationPct}%` : null,
  ].filter(Boolean);

  const definitions = {
    downward_trend: {
      on: run >= t.declineWeeks,
      summary: run >= t.declineWeeks
        ? `The weighted score has fallen ${run} weeks running.`
        : `The score has fallen ${run} week${run === 1 ? '' : 's'} running; the threshold is ${t.declineWeeks}.`,
      evidence: scored.slice(-6).map((w) => ({ label: `Week ${w.week_number}`, value: String(round1(w.raw)) })),
    },
    missed_engagement: {
      on: reasons.length > 0,
      summary: reasons.length ? `Triggered by ${reasons.join('; ')}.` : 'Attendance, homework and participation are all above their thresholds.',
      evidence: [
        { label: 'Assignments not handed in', value: `${missed} of ${setCount}` },
        { label: 'Attendance to date', value: attendancePct == null ? '—' : `${round1(attendancePct)}%` },
        { label: 'Participation to date', value: participationPct == null ? '—' : `${round1(participationPct)}%` },
      ],
    },
    low_projected_grade: {
      on: projected != null && projected < t.projectedGrade,
      summary: projected == null
        ? 'No scores yet, so no projection.'
        : `The score to date averages ${round1(projected)}, against a pass threshold of ${t.projectedGrade}.`,
      evidence: [
        { label: 'Weeks counted', value: String(scored.length) },
        { label: 'Projected grade', value: projected == null ? '—' : String(round1(projected)) },
        { label: 'Pass threshold', value: String(t.projectedGrade) },
      ],
    },
    heavy_commitments: {
      on: hours >= t.commitmentHours,
      summary: hours >= t.commitmentHours
        ? `${round1(hours)} weighted hours of work and care a week, at or above the ${t.commitmentHours}-hour threshold.`
        : `${round1(hours)} weighted hours a week, under the ${t.commitmentHours}-hour threshold.`,
      evidence: [
        { label: 'Weighted hours this week', value: String(round1(hours)) },
        { label: 'Threshold', value: `${t.commitmentHours} hours` },
      ],
    },
    missing_data: {
      on: gap >= t.missingWeeks,
      summary: gap >= t.missingWeeks
        ? `${gap} consecutive weeks with no entries at all.`
        : `The longest gap is ${gap} week${gap === 1 ? '' : 's'}; the threshold is ${t.missingWeeks}.`,
      evidence: weekly.slice(-6).map((w) => ({ label: `Week ${w.week_number}`, value: w.hasEntries ? 'recorded' : 'no entries' })),
    },
  };

  const signals = SIGNAL_KEYS.map((key) => ({
    key,
    label: SIGNAL_LABELS[key],
    active: active[key] !== false,
    on: active[key] !== false && definitions[key].on,
    summary: definitions[key].summary,
    evidence: definitions[key].evidence,
  }));

  const on = signals.filter((s) => s.on);
  return {
    signals,
    signals_on: on.map((s) => s.key),
    level: levelFor(on.length),
    /** So a borderline student is visible rather than hidden by their level. */
    oneAway: on.length < 3 ? levelFor(on.length + 1) : null,
    metrics: {
      declineRun: run,
      missedAssignments: missed,
      attendancePct: round1(attendancePct),
      participationPct: round1(participationPct),
      projected: round1(projected),
      hours: round1(hours),
      missingRun: gap,
    },
  };
}

/** Compare two evaluations to say what changed since the instructor last looked. */
export function diffState(prior, current) {
  const was = prior ? prior.signals_on.length > 0 : false;
  const is = current.signals_on.length > 0;
  if (is && !was) return 'new';
  if (is && was) return 'still';
  if (!is && was) return 'cleared';
  return 'quiet';
}
